addons/sale_stock/models/sale_order.py

```diff
@@ -98,6 +99,14 @@ def _select_expected_date(self, expected_dates):
             return super()._select_expected_date(expected_dates)
         return max(expected_dates)
 
+    @api.constrains('warehouse_id', 'state', 'order_line')
+    def _check_warehouse(self):
+        """ Ensure that the warehouse is set in case of storable products """
+        orders_without_wh = self.filtered(lambda order: order.state not in ('draft', 'cancel') and not order.warehouse_id)
+        goods_so_lines = orders_without_wh.order_line.filtered(lambda line: line.product_id.type == 'consu')
+        if goods_so_lines:
```

Comment:
```suggestion
        if any(l for l in orders_without_wh.order_line if l.product_id.type == 'consu')
```
will not browse the whole list in case of one 'goods' line